front/lib/api/resource_wrappers.ts

```diff
@@ -56,57 +66,129 @@ export function withResourceFetchingFromRoute<
   T,
   U extends ResourceKey,
   A extends SessionOrKeyAuthType,
->(handler: ResourceHandler<T, U, A>, resource: U) {
-  return resolver[resource](handler);
+>(
+  handler: (
+    req: NextApiRequest,
+    res: NextApiResponse<WithAPIErrorResponse<T>>,
+    auth: Authenticator,
+    resources: ResourceMap<U>,
+    sessionOrKeyAuth: A
+  ) => Promise<void> | void,
+  options: OptionsMap<U>
+): (
+  req: NextApiRequest,
+  res: NextApiResponse<WithAPIErrorResponse<T>>,
+  auth: Authenticator,
+  sessionOrKeyAuth: A
+) => Promise<void> | void {
+  const wrappedHandler = resolvers.reduce(
+    (acc, resolver) => resolver(acc),
+    (
+      req: NextApiRequest,
+      res: NextApiResponse<WithAPIErrorResponse<T>>,
+      auth: Authenticator,
+      resources: Partial<ResourceMap<ResourceKey>>,
+      _: Partial<OptionsMap<ResourceKey>>,
+      sessionOrKeyAuth: A
+    ) => {
+      const keys = Object.keys(options) as ResourceKey[];
+      if (!isResourceMap<U>(resources, keys)) {
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message: "Invalid parameters.",
+          },
+        });
+      }
+      return handler(
+        req,
+        res,
+        auth,
+        resources as ResourceMap<U>,
+        sessionOrKeyAuth
+      );
+    }
+  );
+
+  return (
+    req: NextApiRequest,
+    res: NextApiResponse<WithAPIErrorResponse<T>>,
+    auth: Authenticator,
+    sessionOrKeyAuth: A
+  ) => wrappedHandler(req, res, auth, {}, options, sessionOrKeyAuth);
 }
 
 /**
  *  for /w/[wId]/spaces/[spaceId]/... => check the space exists, that it's
  *  not a conversation space, etc. and provide the space resource to the handler.
  */
 function withSpaceFromRoute<T, A extends SessionOrKeyAuthType>(
-  handler: ResourceHandler<T, "space", A>
-) {
+  handler: ResourceHandler<T, A>
+): ResourceHandler<T, A> {
   return async (
     req: NextApiRequest,
     res: NextApiResponse<WithAPIErrorResponse<T>>,
     auth: Authenticator,
+    resources: Partial<ResourceMap<ResourceKey>>,
+    options: Partial<OptionsMap<ResourceKey>>,
     sessionOrKeyAuth: A
   ) => {
     const { spaceId } = req.query;
 
-    // Handling the case where `spaceId` is undefined to keep support for the
-    // legacy endpoint for v1 routes (global space assumed in that case).
-    const shouldKeepLegacyEndpointSupport =
-      sessionOrKeyAuth === null || sessionOrKeyAuth instanceof Authenticator;
-
-    if (typeof spaceId !== "string" && !shouldKeepLegacyEndpointSupport) {
-      return apiError(req, res, {
-        status_code: 400,
-        api_error: {
-          type: "invalid_request_error",
-          message: "Invalid space id.",
-        },
-      });
-    }
+    if (spaceId) {
```

Comment:
Can we fail hard if spaceId is not set and a space has been requested :pray: Same for other resolvers

Reply:
We still have legacy routes without spaceId, where space is deduced from dataSource, so we don't want to fail here. In that case space resolution is handled in `withDataSourceFromRoute` resolver, which check the auth key to see if it can take the system or global space. Space resource will be set with this resolver.

We rather fail in the end if no resolver was able to find required resources (in `front/lib/api/resource_wrappers.ts#95`)

Reply:
ok